File: 2020-11/976.js
/**
 * @param {number[]} A
 * @return {number}
 */
var largestPerimeter = function (A) {
  const n = A.length;

  if (n === 0) return 0;

  A.sort((a, b) => b - a);

  let ans = 0;

  for (let i = 0; i < n; i++) {
    let l = i + 1,
      r = n - 1;
    while (l < r && A[i] >= A[l] + A[r]) r--;
    while (l < r) ans = Math.max(ans, A[i] + A[l] + A[r--]);
  }

  return ans;
};

File: 2020-11/976.cpp
class Solution
{
public:
  int largestPerimeter(vector<int> &A)
  {
    int n = A.size();

    if (n == 0)
      return 0;

    int ans = 0;
    sort(A.begin(), A.end(), [](const int &a, const int &b) {
      return a > b;
    });

    for (int i = 0; i < n - 2; i++)
    {
      if (A[i] < A[i + 1] + A[i + 2])
        return A[i] + A[i + 1] + A[i + 2];
    }
    return ans;
  }
};